test/e2e_federation/hpa.go

```diff
@@ -0,0 +1,605 @@
+				createTargetObjOrFail(f.FederationClientset, dep)
+				hpa := newHpa(nsName, testHpaPrefix, dep.Name, NewInt32(1), NewInt32(50), 1)
+				createHpaOrFail(f.FederationClientset, hpa)
+				waitForClusterListOnTargetObjOrFail(f.FederationClientset, nsName, dep.Name, clusterListLen)
+			})
+		})
+	})
+
+	Describe("Federated Autoscaling: ", func() {
+		var (
+			clusters       fedframework.ClusterSlice
+			nsName         string
+			loadGenPodName string
+		)
+
+		BeforeEach(func() {
+			fedframework.SkipUnlessFederated(f.ClientSet)
+			clusters = f.GetRegisteredClusters()
+			nsName = f.FederationNamespace.Name
+			loadGenPodName = ""
+		})
+
+		AfterEach(func() {
+			fedframework.SkipUnlessFederated(f.ClientSet)
+			// Delete all hpa objects.
+			deleteAllHpasOrFail(f.FederationClientset, nsName)
+		})
+
+		It("given that sufficient replicas are there, hpa replicas get distributed to all clusters", func() {
+			expected, max := getExpectedForHpaWithMaxReplicas(clusters, int32(1), int32(1), int32(1))
+			hpa := createHpaOrFail(f.FederationClientset, newHpa(nsName, testHpaPrefix, testDeploymentPrefix, NewInt32(1), NewInt32(50), max))
+			By(fmt.Sprintf("Creation of hpa %q in namespace %q succeeded.", hpa.Name, nsName))
+			waitForHpaOrFail(f.FederationClientset, f.FederationNamespace.Name, hpa.Name, clusters, expected)
+		})
+
+		Describe("With valid and existing hpa target object: ", func() {
+			AfterEach(func() {
+				fedframework.SkipUnlessFederated(f.ClientSet)
+				// Delete all target objects.
+				deleteAllTargetObjsOrFail(f.FederationClientset, nsName)
+				for _, cluster := range clusters {
+					deleteLoadGenPodOrFail(cluster, loadGenPodName, nsName)
+				}
+			})
+
+			It("fed hpa controls target object", func() {
+				fedframework.SkipUnlessFederated(f.ClientSet)
+
+				// create the deployment object with only 1 replica to ensure
+				// that its created in only 1 cluster, if hpa exists in more clusters, then
+				// it ensures that target object also gets created in more clusters (with 0
+				// replicas where the target object controller could not give replicas).
+				dep := createTargetObjOrFail(f.FederationClientset, newHpaTargetObj(nsName, testDeploymentPrefix, NewInt32(1), ""))
+				expected, max := getExpectedForHpaWithMaxReplicas(clusters, int32(1), int32(1), int32(1))
+				hpa := createHpaOrFail(f.FederationClientset, newHpa(nsName, testHpaPrefix, dep.Name, NewInt32(1), NewInt32(50), max))
+				By(fmt.Sprintf("Creation of hpa %q in namespace %q succeeded.", hpa.Name, nsName))
+
+				waitForHpaOrFail(f.FederationClientset, nsName, hpa.Name, clusters, expected)
+				waitForTargetObjInClustersOrFail(nsName, dep.Name, clusters, expected, false)
+			})
+
+			// The below test does following stuff:
+			// Creates the target deployment object with 2 replica to ensure that both clusters
+			// get 1 replica each.
+			// Create an hpa with min|max : 1|4, to cap the limits during test.
```

Comment:
It seems that max in the code below is 6?  Please check this comment against the code and make sure that they are consistent.